zebrad/src/application.rs

```diff
@@ -503,25 +452,34 @@ impl Application for ZebradApp {
         let _ = stdout().lock().flush();
         let _ = stderr().lock().flush();
 
-        if let Err(e) = self.state().components.shutdown(self, shutdown) {
+        if let Err(e) = self.state().components().shutdown(self, shutdown) {
             let app_name = self.name().to_string();
 
-            // Swap out a fake app so we can trigger the destructor on the original
-            let _ = std::mem::take(self);
+            self.state()
+                .components_mut()
+                .get_downcast_mut::<Tracing>()
+                .map(Tracing::shutdown);
+
             fatal_error(app_name, &e);
         }
-
-        // Swap out a fake app so we can trigger the destructor on the original
-        let _ = std::mem::take(self);
+        self.state()
+            .components_mut()
+            .get_downcast_mut::<Tracing>()
+            .map(Tracing::shutdown);
 
         match shutdown {
             Shutdown::Graceful => process::exit(0),
             Shutdown::Forced => process::exit(1),
             Shutdown::Crash => process::exit(2),
         }
     }
+}
 
-    fn version(&self) -> Version {
-        app_version()
-    }
+/// Boot the given application, parsing subcommand and options from
+/// command-line arguments, and terminating when complete.
+pub fn boot(app_cell: &'static AppCell<ZebradApp>) -> ! {
```

Comment:
```suggestion
// <https://docs.rs/abscissa_core/latest/src/abscissa_core/application.rs.html#174-178>
pub fn boot(app_cell: &'static AppCell<ZebradApp>) -> ! {
```

Reply:
Nitpick: these line numbers could go out of date as the latest abscissa version changes. We could link to the specific version we copied the code from, or name the specific function that people should be looking for in the comment.

Reply:
Fixed in https://github.com/ZcashFoundation/zebra/commit/6b553d290759d64255389c43c1dfb6698313b28b.